File: python/tests/test_m79_genomic_enums.py
"""v0.11 M79 — Modality abstraction + genomic enumerations.

Purely additive groundwork for the genomic milestone series:

* ``Precision.UINT8`` round-trips through every storage provider.
* New :class:`Compression` values 4-8 (rANS / base-pack / quality-binned
  / name-tokenized) persist as HDF5 integer attributes.
* :class:`AcquisitionMode` gains ``GENOMIC_WGS = 7`` and
  ``GENOMIC_WES = 8``.
* Transport :class:`AccessUnit` accepts ``spectrum_class = 5``
  (GenomicRead) without crashing — the wire format is generic.
* Run groups carry an optional ``@modality`` UTF-8 attribute; absent
  attribute → ``"mass_spectrometry"`` for v0.10 backward-compat.
"""
from __future__ import annotations

from pathlib import Path

import numpy as np
import pytest

from ttio.acquisition_run import AcquisitionRun
from ttio.enums import AcquisitionMode, Compression, Precision
from ttio.providers import open_provider
from ttio.providers.hdf5 import Hdf5Provider
from ttio.providers.memory import MemoryProvider
from ttio.providers.sqlite import SqliteProvider
from ttio.spectral_dataset import SpectralDataset, WrittenRun
from ttio.transport.packets import (
    SPECTRUM_CLASS_GENOMIC_READ,
    AccessUnit,
)


# ── UINT8 provider round-trip ────────────────────────────────────────


def _provider_url(provider: str, tmp_path: Path) -> str:
    if provider == "hdf5":
        return str(tmp_path / "m79.h5")
    if provider == "memory":
        return f"memory://m79-{provider}-{id(tmp_path)}"
    if provider == "sqlite":
        return str(tmp_path / "m79.tio.sqlite")
    if provider == "zarr":
        return f"zarr://{tmp_path / 'm79.zarr'}"
    raise ValueError(provider)


def _discard_memory(provider: str, url: str) -> None:
    if provider == "memory":
        MemoryProvider.discard_store(url)


@pytest.mark.parametrize("provider", ["hdf5", "memory", "sqlite", "zarr"])
def test_uint8_roundtrip(provider: str, tmp_path: Path) -> None:
    """1000-element UINT8 buffer round-trips byte-exactly."""
    if provider == "zarr":
        pytest.importorskip("zarr", minversion="3.0")

    url = _provider_url(provider, tmp_path)
    expected = np.arange(1000, dtype=np.uint8)  # 0..255 repeating

    try:
        with open_provider(url, provider=provider, mode="w") as p:
            ds = p.root_group().create_dataset(
                "bases", Precision.UINT8, length=len(expected))
            ds.write(expected)
            assert ds.precision == Precision.UINT8

        with open_provider(url, provider=provider, mode="r") as p:
            got = p.root_group().open_dataset("bases").read()
            assert got.dtype == np.uint8
            np.testing.assert_array_equal(got, expected)
    finally:
        _discard_memory(provider, url)


@pytest.mark.parametrize("provider", ["hdf5", "memory", "sqlite", "zarr"])
def test_uint8_partial_read(provider: str, tmp_path: Path) -> None:
    """Hyperslab read of UINT8 returns the requested window."""
    if provider == "zarr":
        pytest.importorskip("zarr", minversion="3.0")

    url = _provider_url(provider, tmp_path)
    expected = np.arange(1000, dtype=np.uint8)

    try:
        with open_provider(url, provider=provider, mode="w") as p:
            ds = p.root_group().create_dataset(
                "bases", Precision.UINT8, length=len(expected))
            ds.write(expected)

        with open_provider(url, provider=provider, mode="r") as p:
            got = p.root_group().open_dataset("bases").read(
                offset=500, count=100)
            np.testing.assert_array_equal(got, expected[500:600])
    finally:
        _discard_memory(provider, url)


# ── Compression enum persistence ─────────────────────────────────────


def test_compression_enum_values() -> None:
    """All compression IDs in v1.0. On-disk integers — must not drift.
    Slots 8 (NAME_TOKENIZED v1), 9 (REF_DIFF v1), and 10
    (FQZCOMP_NX16 v1) are intentionally absent — all three legacy v1
    codecs were removed in the v1.0 reset (Phases 2c/2d). Their v2
    successors live at 13/14/15."""
    # M79 (v0.11) — base codec set
    assert Compression.RANS_ORDER0.value == 4
    assert Compression.RANS_ORDER1.value == 5
    assert Compression.BASE_PACK.value == 6
    assert Compression.QUALITY_BINNED.value == 7
    # 8/9/10 reserved (v1 codecs removed in v1.0 reset)
    assert Compression.DELTA_RANS_ORDER0.value == 11
    assert Compression.FQZCOMP_NX16_Z.value == 12
    # v1.7+ #11 channels
    assert Compression.MATE_INLINE_V2.value == 13      # v1.7 ch1
    assert Compression.REF_DIFF_V2.value == 14         # v1.8 ch2
    assert Compression.NAME_TOKENIZED_V2.value == 15   # v1.9 ch3
    # IDs 8, 9, 10 must remain unmapped
    import pytest as _pytest
    for unmapped in (8, 9, 10):
        with _pytest.raises(ValueError):
            Compression(unmapped)


@pytest.mark.parametrize("codec", [
    Compression.RANS_ORDER0,
    Compression.RANS_ORDER1,
    Compression.BASE_PACK,
    Compression.QUALITY_BINNED,
    Compression.DELTA_RANS_ORDER0,
    Compression.FQZCOMP_NX16_Z,
    Compression.MATE_INLINE_V2,
    Compression.REF_DIFF_V2,
    Compression.NAME_TOKENIZED_V2,
])
def test_compression_enum_persists_as_attribute(
        codec: Compression, tmp_path: Path) -> None:
    """Each new codec ID round-trips as an HDF5 integer attribute."""
    path = str(tmp_path / "codec.h5")
    with Hdf5Provider.open(path, mode="w") as p:
        p.root_group().set_attribute("codec_id", int(codec.value))

    with Hdf5Provider.open(path, mode="r") as p:
        raw = p.root_group().get_attribute("codec_id")
        assert int(raw) == codec.value
        assert Compression(int(raw)) is codec


# ── AcquisitionMode genomic values ──────────────────────────────────


def test_acquisition_mode_genomic_values() -> None:
    """Stable wire integers for genomic acquisition modes."""
    assert AcquisitionMode.GENOMIC_WGS.value == 7
    assert AcquisitionMode.GENOMIC_WES.value == 8


# ── Transport spectrum_class = 5 ────────────────────────────────────


def test_access_unit_spectrum_class_genomic_read_roundtrip() -> None:
    """Codec must serialise + deserialise spectrum_class=5 without
    crashing. Genomic AUs reuse the spectral prefix with zeroed
    spectral fields — proper genomic prefix arrives in M82."""
    au = AccessUnit(
        spectrum_class=SPECTRUM_CLASS_GENOMIC_READ,
        acquisition_mode=int(AcquisitionMode.GENOMIC_WGS),
        ms_level=0,
        polarity=2,  # unknown
        retention_time=0.0,
        precursor_mz=0.0,
        precursor_charge=0,
        ion_mobility=0.0,
        base_peak_intensity=0.0,
        channels=[],
    )

    blob = au.to_bytes()
    decoded = AccessUnit.from_bytes(blob)

    assert decoded.spectrum_class == SPECTRUM_CLASS_GENOMIC_READ == 5
    assert decoded.acquisition_mode == int(AcquisitionMode.GENOMIC_WGS)
    assert decoded.channels == []
    assert decoded.pixel_x == 0  # MSImagePixel extension MUST NOT activate


# ── Modality attribute (read-side; write-side lands with M74) ───────


def _write_minimal_run(path: Path) -> None:
    """Smallest valid v1.1 .tio with one MS run for modality tests."""
    n_spec, n_pts = 2, 4
    offsets = np.arange(n_spec, dtype=np.uint64) * n_pts
    lengths = np.full(n_spec, n_pts, dtype=np.uint32)
    mz = np.tile(np.linspace(100.0, 200.0, n_pts), n_spec).astype(np.float64)
    intensity = np.tile(
        np.linspace(1.0, 1000.0, n_pts), n_spec).astype(np.float64)
    run = WrittenRun(
        spectrum_class="TTIOMassSpectrum",
        acquisition_mode=int(AcquisitionMode.MS1_DDA),
        channel_data={"mz": mz, "intensity": intensity},
        offsets=offsets,
        lengths=lengths,
        retention_times=np.linspace(0.0, 1.0, n_spec, dtype=np.float64),
        ms_levels=np.ones(n_spec, dtype=np.int32),
        polarities=np.ones(n_spec, dtype=np.int32),
        precursor_mzs=np.zeros(n_spec, dtype=np.float64),
        precursor_charges=np.zeros(n_spec, dtype=np.int32),
        base_peak_intensities=np.full(n_spec, 1000.0, dtype=np.float64),
    )
    SpectralDataset.write_minimal(
        path,
        title="m79 modality",
        isa_investigation_id="TTIO:m79",
        runs={"run_0001": run},
    )


def test_modality_default_for_v0_10_files(tmp_path: Path) -> None:
    """Files written without a ``@modality`` attribute (= every v0.10
    file) read as ``"mass_spectrometry"`` so existing readers don't
    break on the new field."""
    out = tmp_path / "default.tio"
    _write_minimal_run(out)

    with SpectralDataset.open(out) as ds:
        run = ds.ms_runs["run_0001"]
        assert run.modality == "mass_spectrometry"


def test_modality_explicit_genomic_sequencing(tmp_path: Path) -> None:
    """An explicit ``@modality = "genomic_sequencing"`` set on the run
    group is preserved through a close + reopen cycle. M79 ships only
    the read side; the write here goes through the StorageGroup
    attribute API directly so we don't need GenomicRun.write_to_group
    (M74)."""
    out = tmp_path / "explicit.tio"
    _write_minimal_run(out)

    # Re-open in read-write mode and stamp the attribute. SpectralDataset
    # doesn't surface a writable run group directly, so reach in via the
    # provider.
    with Hdf5Provider.open(str(out), mode="r+") as prov:
        run_group = (prov.root_group()
                         .open_group("study")
                         .open_group("ms_runs")
                         .open_group("run_0001"))
        run_group.set_attribute("modality", "genomic_sequencing")

    with SpectralDataset.open(out) as ds:
        run = ds.ms_runs["run_0001"]
        assert run.modality == "genomic_sequencing"

File: objc/Tests/TestM79GenomicEnums.m
// TestM79GenomicEnums.m — v0.11 M79.
//
// Modality abstraction + genomic enumerations. Purely additive
// groundwork for the v0.11 genomic milestone series:
//
//   * TTIOPrecisionUInt8 round-trips through HDF5 + Memory providers.
//   * New TTIOCompression ordinals 4-8 (rANS / base-pack /
//     quality-binned / name-tokenized) have stable integer values.
//   * TTIOAcquisitionMode gains GenomicWGS = 7, GenomicWES = 8.
//   * Transport TTIOAccessUnit accepts spectrumClass = 5 (GenomicRead)
//     without crashing; MSImagePixel extension MUST NOT activate.
//   * TTIOAcquisitionRun.modality defaults to @"mass_spectrometry"
//     so v0.10 files surface as mass-spec runs.
//
// SPDX-License-Identifier: LGPL-3.0-or-later

#import <Foundation/Foundation.h>
#import "Testing.h"
#import "Providers/TTIOStorageProtocols.h"
#import "Providers/TTIOProviderRegistry.h"
#import "Providers/TTIOMemoryProvider.h"
#import "Transport/TTIOAccessUnit.h"
#import "Run/TTIOAcquisitionRun.h"
#import "Run/TTIOInstrumentConfig.h"
#import "ValueClasses/TTIOEncodingSpec.h"
#import "ValueClasses/TTIOEnums.h"
#import <unistd.h>

static NSString *m79GePath(NSString *suffix)
{
    return [NSString stringWithFormat:@"/tmp/ttio_m79ge_%d_%@",
            (int)getpid(), suffix];
}

static NSData *uint8Sample(NSUInteger n)
{
    NSMutableData *d = [NSMutableData dataWithLength:n];
    uint8_t *p = (uint8_t *)d.mutableBytes;
    for (NSUInteger i = 0; i < n; i++) {
        p[i] = (uint8_t)(i & 0xFF);
    }
    return d;
}

static void uint8RoundTripThroughProvider(NSString *providerName,
                                           NSString *url)
{
    const NSUInteger N = 1000;
    NSData *expected = uint8Sample(N);
    NSError *err = nil;

    id<TTIOStorageProvider> w =
        [[TTIOProviderRegistry sharedRegistry]
            openURL:url
               mode:TTIOStorageOpenModeCreate
           provider:providerName
              error:&err];
    PASS(w != nil, "M79: provider opens for create (%s)",
         [providerName UTF8String]);

    id<TTIOStorageGroup> root = [w rootGroupWithError:&err];
    id<TTIOStorageDataset> ds =
        [root createDatasetNamed:@"bases"
                        precision:TTIOPrecisionUInt8
                           length:N
                        chunkSize:0
                      compression:TTIOCompressionNone
                 compressionLevel:0
                            error:&err];
    PASS(ds != nil, "M79: UINT8 dataset created (%s)",
         [providerName UTF8String]);
    PASS([ds writeAll:expected error:&err],
         "M79: UINT8 dataset writes (%s)", [providerName UTF8String]);
    [w close];

    id<TTIOStorageProvider> r =
        [[TTIOProviderRegistry sharedRegistry]
            openURL:url
               mode:TTIOStorageOpenModeRead
           provider:providerName
              error:&err];
    PASS(r != nil, "M79: provider reopens read-only (%s)",
         [providerName UTF8String]);
    id<TTIOStorageGroup> root2 = [r rootGroupWithError:&err];
    id<TTIOStorageDataset> back =
        [root2 openDatasetNamed:@"bases" error:&err];
    PASS(back != nil, "M79: UINT8 dataset reopens (%s)",
         [providerName UTF8String]);
    PASS([back precision] == TTIOPrecisionUInt8,
         "M79: precision survives round-trip (%s)",
         [providerName UTF8String]);

    NSData *got = [back readAll:&err];
    PASS(got != nil, "M79: UINT8 readAll returned data (%s)",
         [providerName UTF8String]);
    PASS([got isEqualToData:expected],
         "M79: UINT8 bytes round-trip byte-exactly (%s)",
         [providerName UTF8String]);
    [r close];
}

void testM79GenomicEnums(void)
{
    // ── 1. UINT8 round-trip via HDF5 provider ──────────────────────
    {
        NSString *path = m79GePath(@"uint8.h5");
        unlink([path fileSystemRepresentation]);
        uint8RoundTripThroughProvider(@"hdf5", path);
        unlink([path fileSystemRepresentation]);
    }

    // ── 2. UINT8 round-trip via Memory provider ────────────────────
    {
        NSString *url = [NSString stringWithFormat:@"memory://m79ge-%d",
                         (int)getpid()];
        [TTIOMemoryProvider discardStore:url];
        uint8RoundTripThroughProvider(@"memory", url);
        [TTIOMemoryProvider discardStore:url];
    }

    // ── 3. TTIOPrecisionUInt8 element size == 1 ────────────────────
    {
        TTIOEncodingSpec *spec =
            [TTIOEncodingSpec specWithPrecision:TTIOPrecisionUInt8
                           compressionAlgorithm:TTIOCompressionNone
                                      byteOrder:TTIOByteOrderLittleEndian];
        PASS([spec elementSize] == 1,
             "M79: TTIOPrecisionUInt8 element size is 1 byte");
    }

    // ── 4. New compression ordinals are stable ─────────────────────
    {
        PASS((NSUInteger)TTIOCompressionNone             == 0, "Compression.None == 0");
        PASS((NSUInteger)TTIOCompressionZlib             == 1, "Compression.Zlib == 1");
        PASS((NSUInteger)TTIOCompressionLZ4              == 2, "Compression.LZ4 == 2");
        PASS((NSUInteger)TTIOCompressionNumpressDelta    == 3, "Compression.NumpressDelta == 3");
        PASS((NSUInteger)TTIOCompressionRansOrder0       == 4, "Compression.RansOrder0 == 4");
        PASS((NSUInteger)TTIOCompressionRansOrder1       == 5, "Compression.RansOrder1 == 5");
        PASS((NSUInteger)TTIOCompressionBasePack         == 6, "Compression.BasePack == 6");
        PASS((NSUInteger)TTIOCompressionQualityBinned    == 7, "Compression.QualityBinned == 7");
        // Slots 8/9/10 reserved (v1 codecs removed in v1.0 reset).
        PASS((NSUInteger)TTIOCompressionDeltaRansOrder0  == 11, "Compression.DeltaRansOrder0 == 11");
        PASS((NSUInteger)TTIOCompressionFqzcompNx16Z     == 12, "Compression.FqzcompNx16Z == 12");
        PASS((NSUInteger)TTIOCompressionMateInlineV2     == 13, "Compression.MateInlineV2 == 13");
        PASS((NSUInteger)TTIOCompressionRefDiffV2        == 14, "Compression.RefDiffV2 == 14");
        PASS((NSUInteger)TTIOCompressionNameTokenizedV2  == 15, "Compression.NameTokenizedV2 == 15");
    }

    // ── 5. AcquisitionMode genomic ordinals are stable ─────────────
    {
        PASS((NSUInteger)TTIOAcquisitionModeMS1DDA      == 0, "AcquisitionMode.MS1DDA == 0");
        PASS((NSUInteger)TTIOAcquisitionModeGenomicWGS  == 7, "AcquisitionMode.GenomicWGS == 7");
        PASS((NSUInteger)TTIOAcquisitionModeGenomicWES  == 8, "AcquisitionMode.GenomicWES == 8");
    }

    // ── 6. AccessUnit spectrumClass=5 (GenomicRead) round-trip ─────
    {
        TTIOAccessUnit *au =
            [[TTIOAccessUnit alloc]
                initWithSpectrumClass:5
                      acquisitionMode:(uint8_t)TTIOAcquisitionModeGenomicWGS
                              msLevel:0
                             polarity:2     // unknown — wire convention
                        retentionTime:0.0
                          precursorMz:0.0
                      precursorCharge:0
                          ionMobility:0.0
                    basePeakIntensity:0.0
                             channels:@[]
                               pixelX:0
                               pixelY:0
                               pixelZ:0];
        PASS(au != nil, "M79: GenomicRead AU constructible");

        NSData *blob = [au encode];
        PASS(blob != nil && blob.length >= 38,
             "M79: GenomicRead AU encodes (>= 38 byte fixed prefix)");

        NSError *err = nil;
        TTIOAccessUnit *back =
            [TTIOAccessUnit decodeFromBytes:(const uint8_t *)blob.bytes
                                     length:blob.length
                                      error:&err];
        PASS(back != nil, "M79: GenomicRead AU decodes");
        PASS(back.spectrumClass == 5, "M79: spectrumClass preserved as 5");
        PASS(back.acquisitionMode ==
                 (uint8_t)TTIOAcquisitionModeGenomicWGS,
             "M79: acquisitionMode preserved as GenomicWGS");
        PASS(back.channels.count == 0, "M79: zero channels preserved");
        // MSImagePixel extension MUST NOT activate for spectrumClass=5.
        PASS(back.pixelX == 0, "M79: pixelX stays 0 for GenomicRead");
        PASS(back.pixelY == 0, "M79: pixelY stays 0 for GenomicRead");
        PASS(back.pixelZ == 0, "M79: pixelZ stays 0 for GenomicRead");
    }

    // ── 7. AcquisitionRun.modality default ─────────────────────────
    {
        TTIOInstrumentConfig *cfg =
            [[TTIOInstrumentConfig alloc]
                initWithManufacturer:@"Test"
                               model:@"M79"
                        serialNumber:@""
                          sourceType:@""
                        analyzerType:@""
                        detectorType:@""];
        TTIOAcquisitionRun *run =
            [[TTIOAcquisitionRun alloc]
                initWithSpectra:@[]
                acquisitionMode:TTIOAcquisitionModeMS1DDA
               instrumentConfig:cfg];
        PASS(run != nil, "M79: empty MS run constructible");
        PASS([run.modality isEqualToString:@"mass_spectrometry"],
             "M79: modality defaults to mass_spectrometry");
    }
}
